### ISSUE TITLE
[BUG - Develop] Global Scripts loading doesn't catch parse errors

### ISSUE THREAD

**Issue by InspectorCaracal** [evennia/evennia#2706]
#### Describe the bug
When there's an error with loading a global scripts module, e.g. it imports the `Room` class but that class has a syntax error, the loading error isn't caught and the traceback points to [this line](https://github.com/evennia/evennia/blob/251a70275bbd98a3e157cbb4c025597a4bb24ac9/evennia/utils/containers.py#L135) with a KeyError

#### To Reproduce
1. Create a simple global script which imports a class such as `Room` from your game dir, and add it to `settings.py`
2. Introduce a syntax error to your `Room` class, such as an extra space in an indent.
3. Reload the server.

#### Expected behavior
It'd be more useful if there was an error thrown which indicated that the script had failed being loaded (and, ideally, why), as the current error invariably leads people to think there is a problem with the global scripts dict in `settings.py`.

#### Develop-branch commit
f093c8bcb

#### Additional context
I feel like it should be catching that kind of an error in [`class_from_module`](https://github.com/evennia/evennia/blob/f093c8bcb1321a44e33003cf91cf9f565e028de7/evennia/utils/utils.py#L1524) or it should at least be throwing an exception in [`load_data`](https://github.com/evennia/evennia/blob/f093c8bcb1321a44e33003cf91cf9f565e028de7/evennia/utils/containers.py#L202) but neither seem to be doing so. Instead, they just silently don't load the script.

**Comment by Henddher** [comment]
Hi @InspectorCaracal

Thanks for the steps to reproduce!

This is the stack trace I'm getting. I think it matches your description.

```shell
...evennia_mygame$ evennia start
Traceback (most recent call last):
  File "/home/henddher/Documents/github/evennia_venv/bin/evennia", line 10, in <module>
    main()
  File "/home/henddher/Documents/github/evennia/evennia/server/evennia_launcher.py", line 2273, in main
    init_game_directory(CURRENT_DIR, check_db=True)
  File "/home/henddher/Documents/github/evennia/evennia/server/evennia_launcher.py", line 1761, in init_game_directory
    check_database()
  File "/home/henddher/Documents/github/evennia/evennia/server/evennia_launcher.py", line 1462, in check_database
    evennia._init()
  File "/home/henddher/Documents/github/evennia/evennia/__init__.py", line 363, in _init
    GLOBAL_SCRIPTS.start()
  File "/home/henddher/Documents/github/evennia/evennia/utils/containers.py", line 188, in start
    self._load_script(key)
  File "/home/henddher/Documents/github/evennia/evennia/utils/containers.py", line 136, in _load_script
    typeclass = self.typeclass_storage[key]
KeyError: 'key'

```
I also agree with your assessment. 

It would make sense to `try/catch` around loading each script defined in `GLOBAL_SCRIPTS` within `mygame/server/conf/settings.py`. That way, should the `typeclass` (or any other *important* key) had issues, the server would not fail to start. On the other hand, given that these are *global scripts*, maybe failing catastrophically does make sense? one would not want their `typeclasses.godlyscripts.LifeOrDeadScript` not to start. I think it's a bit philosophical (imo).

I'll take a stab at it with `try/catch` and some guards around loading. I am sure @Griatch will have more insights.

One thing I am not sure about is creating a test for it. That might take a bit longer but worth trying.

Thoughts?

**Comment by InspectorCaracal** [comment]
@Henddher That's the one!

**Comment by Henddher** [comment]
At first, I figure that there would be no `try/catch` around those loading functions. But, they are. I think the issue is that the exception is silently swallowed and the system proceeds as normal.

https://github.com/Henddher/evennia/blob/39f6505bf6438eaefbd0791e5c821578b63a3dca/evennia/utils/containers.py#L192-L201
```python
            for key, data in self.loaded_data.items():
                try:
                    typeclass = data.get("typeclass", settings.BASE_SCRIPT_TYPECLASS)
                    self.typeclass_storage[key] = class_from_module(typeclass)
                except ImportError as err:
                    logger.log_err(
                        f"GlobalScriptContainer could not start global script {key}: {err}"
                    )
                    #  👈 No raise?


    def get(self, key, default=None):
```

**Comment by Henddher** [comment]
I've got a solution (I think!).

https://github.com/evennia/evennia/pull/2882/commits/d9bdabf4c183dc9f93986395fb6e836da2a005d8

Instead of simply "swallowing" the exception, I remove `key` from `self.loaded_data` essentially *skipping* the final load of that poorly constructed global script. All other global scripts will be attempted.

This is my `mygame/server/conf/settings.py`

```python
GLOBAL_SCRIPTS = {
    'nothingscript': {
            'typeclass': 'typeclasses.badscript.NothingScript',
            'desc': 'Nothing script'
    },
    'badscript': {
            'typeclass': 'typeclasses.badscript.BadScript',
            'desc': 'Bad script'
    },
    'defaultscript': {
            'desc': 'Default script'
    },
```

And my `mygame/typeclasses/badscript.py`

```python
from evennia import DefaultScript
from evennia import bullcrap

class BaddScript:
  pass

class NothingScript(DefaultScript):
  pass
```

And, when you run `evennia start` (or `evennia stop`), this is what the log contains now:

```text
2022-10-02 02:00:34 [..] Loaded.
2022-10-02 02:00:33 [..] Loading /home/henddher/Documents/github/evennia/evennia/server/server.py...
2022-10-02 02:00:34 [EE] GlobalScriptContainer could not start import global script nothingscript. It will be removed (skipped).
2022-10-02 02:00:34 [EE] GlobalScriptContainer could not start import global script badscript. It will be removed (skipped).
2022-10-02 02:00:34 [!!] Traceback (most recent call last):
2022-10-02 02:00:34 [!!]   File "/home/henddher/Documents/github/evennia/evennia/typeclasses/models.py", line 271, in set_class_from_typeclass
2022-10-02 02:00:34 [!!]     self.__class__ = class_from_module(self.db_typeclass_path)
2022-10-02 02:00:34 [!!]   File "/home/henddher/Documents/github/evennia/evennia/utils/utils.py", line 1569, in class_from_module
2022-10-02 02:00:34 [!!]     mod = importlib.import_module(testpath, package="evennia")
2022-10-02 02:00:34 [!!]   File "/home/henddher/.pyenv/versions/3.10.4/lib/python3.10/importlib/__init__.py", line 126, in import_module
2022-10-02 02:00:34 [!!]     return _bootstrap._gcd_import(name[level:], package, level)
2022-10-02 02:00:34 [!!]   File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
2022-10-02 02:00:34 [!!]   File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
2022-10-02 02:00:34 [!!]   File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
2022-10-02 02:00:34 [!!]   File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
2022-10-02 02:00:34 [!!]   File "<frozen importlib._bootstrap_external>", line 883, in exec_module
2022-10-02 02:00:34 [!!]   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
2022-10-02 02:00:34 [!!]   File "/home/henddher/Documents/github/evennia_mygame/typeclasses/badscript.py", line 2, in <module>
2022-10-02 02:00:34 [!!]     from evennia import bullcrap
2022-10-02 02:00:34 [!!] ImportError: cannot import name 'bullcrap' from 'evennia' (/home/henddher/Documents/github/evennia/evennia/__init__.py)
2022-10-02 02:00:34 [..] Loaded.
2022-10-02 02:00:34 [..] twistd 21.7.0 (/home/henddher/Documents/github/evennia_venv/bin/python 3.10.4) starting up.
2022-10-02 02:00:34 [..] reactor class: twisted.internet.epollreactor.EPollReactor.
2022-10-02 02:00:34 [..] Webserver starting on 4005
2022-10-02 02:00:34 [..] Evennia Server successfully started.
```

**Comment by Henddher** [comment]
Thoughts?

I'll push it for review and maybe push another PR with a test (if I figure out how to test that).

**Comment by InspectorCaracal** [comment]
The error is much clearer, which is a definite improvement, but I'd prefer if the game failed to start due to a broken global script rather than continuing to start without it. If I have a global script, the chances that I have other systems relying on it are _very_ high, so having the game start without that script available could cause a cascade of broken systems.

**Comment by Henddher** [comment]
I would agree, but imo it's a philosophical topic ... 
I advocate for a flag to turn it OFF/ON!

Thoughts?

**Comment by InspectorCaracal** [comment]
I think that's unnecessarily complicated.

If you really want the game to start regardless of a completely broken module, then I would suggest following prior example and having the script use a default fallback typeclass like the other systems.

**Comment by Henddher** [comment]
Then another one liner `raise e` would have worked too :smile:

**Comment by Griatch** [comment]
Closed in #2882.
